src/logic.py

```diff
@@ -182,3 +187,68 @@ def dispatch_css_maintenance_action(maintenance_mode: bool):
             logger.info('CSS GH API status: %s' % x.status_code)
         else:
             logger.error('GH API error: %s' % x.content)
+
+
+def dispatch_uptime_incident(enable_incident: bool):
+    global incident_id
+
+    url = config.get('uptime_status_api')
+    uptime_token = config.get('uptime_status_token')
+    uptime_statuspage_id = config.get('uptime_status_page_id')
+    namespace = config.get('namespace')
+    env = namespace[7:]
+    if uptime_token == "" or uptime_statuspage_id == "":
+        logger.error('The uptime status page incident creation/closure has not been configured')
+        return
+
+    incident_url = f"{url}{uptime_statuspage_id}/incidents/"
+    bearer = 'token %s' % uptime_token
+    headers = {'Authorization': bearer}
+
+    body = {
+        "name": f"[{env}] Keycloak is in Disaster Recovery mode",
+        "include_in_global_metrics": False,
+        "updates": [
+            {
+                "id": 0,
+                "description": f"The switchover to GoldDR was triggered for the {env} environment.",
+                "incident_state": "investigating"
+            }
+        ],
+        "incident_type": "INCIDENT",
+        "update_component_status": True,
+        "notify_subscribers": True
+    }
+
+    if enable_incident:
+        try:
+            x = requests.post(incident_url, json=body, headers=headers)
+            if x.status_code == 200:
+                incident_id = x.json()["results"]["pk"]
+                logger.info(f"Uptime incident {incident_id} was created.")
+            else:
+                logger.error(f"Uptime incident creation returned code {x.status_code}")
+                incident_id = None
+        except BaseException:
+            logger.error("Uptime incident creation failed.")
+            incident_id = None
+    else:
+        # close the incident
+        try:
+            close_body = body
+            close_body["ends_at"] = datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+            close_body["incident_state"] = "resolved"
+            close_body["updates"] = [{
+                "id": 0,
+                "description": f"The {env} environement is no longer in disaster recovery mode.",
+                "incident_state": "resolved"
+            }]
+            x = requests.patch(f"{incident_url}{incident_id}", json=body, headers=headers)
```

Comment:
@thegentlemanphysicist In here if the incident was manually deleted, this step may fail so can we validate if incident exists before updating it?

Reply:
In that case, the patch will restore 404, which is caught by the next line.